### PR TITLE
[CARBONDATA-1359] Unable to use carbondata on hive

### PR DESCRIPTION
**Pull request by shivangi1015**

### CODE REVIEW

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

**Comment by CarbonDataQA** [comment]
Can one of the admins verify this patch?

**Comment by CarbonDataQA** [comment]
Can one of the admins verify this patch?

**Comment by CarbonDataQA** [comment]
Can one of the admins verify this patch?

**Comment by chenliang613** [comment]
1.please correct your PR title as per : https://github.com/apache/carbondata/blob/master/docs/How-to-contribute-to-Apache-CarbonData.md
2.Can you please provide the reproduce steps for this issue.

**Comment by shivangi1015** [comment]
1. I have corrected the PR title.
2. Steps to reproduce:
     i. alter table my_user set FILEFORMAT
       INPUTFORMAT "org.apache.carbondata.hive.MapredCarbonInputFormat"
       OUTPUTFORMAT "org.apache.carbondata.hive.MapredCarbonOutputFormat"
       SERDE "org.apache.carbondata.hive.CarbonHiveSerDe";
    ii. set hive.mapred.supports.subdirectories=true;
    iii. set mapreduce.input.fileinputformat.input.dir.recursive=true;
    iv. alter table my_user set LOCATION 'hdfs://localhost:54310/opt/data/my_user' ;   
    v. select * from my_user;

Result : 
org.apache.hive.service.cli.HiveSQLException: java.io.IOException: java.lang.NullPointerException
at org.apache.hive.service.cli.operation.SQLOperation.getNextRowSet(SQLOperation.java:352)
at org.apache.hive.service.cli.operation.OperationManager.getOperationNextRowSet(OperationManager.java:220)
at org.apache.hive.service.cli.session.HiveSessionImpl.fetchResults(HiveSessionImpl.java:685)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:497)
at org.apache.hive.service.cli.session.HiveSessionProxy.invoke(HiveSessionProxy.java:78)
at org.apache.hive.service.cli.session.HiveSessionProxy.access$000(HiveSessionProxy.java:36)
at org.apache.hive.service.cli.session.HiveSessionProxy$1.run(HiveSessionProxy.java:63)
at java.security.AccessController.doPrivileged(Native Method)
at javax.security.auth.Subject.doAs(Subject.java:422)
at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1491)
at org.apache.hive.service.cli.session.HiveSessionProxy.invoke(HiveSessionProxy.java:59)
at com.sun.proxy.$Proxy32.fetchResults(Unknown Source)
at org.apache.hive.service.cli.CLIService.fetchResults(CLIService.java:454)
at org.apache.hive.service.cli.thrift.ThriftCLIService.FetchResults(ThriftCLIService.java:621)
at org.apache.hive.service.cli.thrift.TCLIService$Processor$FetchResults.getResult(TCLIService.java:1553)
at org.apache.hive.service.cli.thrift.TCLIService$Processor$FetchResults.getResult(TCLIService.java:1538)
at org.apache.thrift.ProcessFunction.process(ProcessFunction.java:39)
at org.apache.thrift.TBaseProcessor.process(TBaseProcessor.java:39)
at org.apache.hive.service.auth.TSetIpAddressProcessor.process(TSetIpAddressProcessor.java:56)
at org.apache.thrift.server.TThreadPoolServer$WorkerProcess.run(TThreadPoolServer.java:285)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: java.lang.NullPointerException
at org.apache.hadoop.hive.ql.exec.FetchOperator.getNextRow(FetchOperator.java:507)
at org.apache.hadoop.hive.ql.exec.FetchOperator.pushRow(FetchOperator.java:414)
at org.apache.hadoop.hive.ql.exec.FetchTask.fetch(FetchTask.java:140)
at org.apache.hadoop.hive.ql.Driver.getResults(Driver.java:1670)
at org.apache.hive.service.cli.operation.SQLOperation.getNextRowSet(SQLOperation.java:347)
... 25 more
Caused by: java.lang.NullPointerException
at org.apache.carbondata.core.metadata.AbsoluteTableIdentifier.fromTablePath(AbsoluteTableIdentifier.java:84)
at org.apache.carbondata.hive.MapredCarbonInputFormat.populateCarbonTable(MapredCarbonInputFormat.java:105)
at org.apache.carbondata.hive.MapredCarbonInputFormat.getCarbonTable(MapredCarbonInputFormat.java:115)
at org.apache.carbondata.hive.MapredCarbonInputFormat.getQueryModel(MapredCarbonInputFormat.java:122)
at org.apache.carbondata.hive.MapredCarbonInputFormat.getRecordReader(MapredCarbonInputFormat.java:76)
at org.apache.hadoop.hive.ql.exec.FetchOperator$FetchInputFormatSplit.getRecordReader(FetchOperator.java:673)
at org.apache.hadoop.hive.ql.exec.FetchOperator.getRecordReader(FetchOperator.java:323)
at org.apache.hadoop.hive.ql.exec.FetchOperator.getNextRow(FetchOperator.java:445)
... 29 more
Exception in thread "main" org.apache.hive.service.cli.HiveSQLException: java.io.IOException: java.lang.NullPointerException
at org.apache.hive.jdbc.Utils.verifySuccess(Utils.java:256)
at org.apache.hive.jdbc.Utils.verifySuccessWithInfo(Utils.java:242)
at org.apache.hive.jdbc.HiveQueryResultSet.next(HiveQueryResultSet.java:365)
at org.apache.carbondata.hiveexample.HiveExample$.main(HiveExample.scala:108)
at org.apache.carbondata.hiveexample.HiveExample.main(HiveExample.scala)
Caused by: org.apache.hive.service.cli.HiveSQLException: java.io.IOException: java.lang.NullPointerException
at org.apache.hive.service.cli.operation.SQLOperation.getNextRowSet(SQLOperation.java:352)
at org.apache.hive.service.cli.operation.OperationManager.getOperationNextRowSet(OperationManager.java:220)
at org.apache.hive.service.cli.session.HiveSessionImpl.fetchResults(HiveSessionImpl.java:685)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:497)
at org.apache.hive.service.cli.session.HiveSessionProxy.invoke(HiveSessionProxy.java:78)
at org.apache.hive.service.cli.session.HiveSessionProxy.access$000(HiveSessionProxy.java:36)
at org.apache.hive.service.cli.session.HiveSessionProxy$1.run(HiveSessionProxy.java:63)
at java.security.AccessController.doPrivileged(Native Method)
at javax.security.auth.Subject.doAs(Subject.java:422)
at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1491)
at org.apache.hive.service.cli.session.HiveSessionProxy.invoke(HiveSessionProxy.java:59)
at com.sun.proxy.$Proxy32.fetchResults(Unknown Source)
at org.apache.hive.service.cli.CLIService.fetchResults(CLIService.java:454)
at org.apache.hive.service.cli.thrift.ThriftCLIService.FetchResults(ThriftCLIService.java:621)
at org.apache.hive.service.cli.thrift.TCLIService$Processor$FetchResults.getResult(TCLIService.java:1553)
at org.apache.hive.service.cli.thrift.TCLIService$Processor$FetchResults.getResult(TCLIService.java:1538)
at org.apache.thrift.ProcessFunction.process(ProcessFunction.java:39)
at org.apache.thrift.TBaseProcessor.process(TBaseProcessor.java:39)
at org.apache.hive.service.auth.TSetIpAddressProcessor.process(TSetIpAddressProcessor.java:56)
at org.apache.thrift.server.TThreadPoolServer$WorkerProcess.run(TThreadPoolServer.java:285)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: java.lang.NullPointerException
at org.apache.hadoop.hive.ql.exec.FetchOperator.getNextRow(FetchOperator.java:507)
at org.apache.hadoop.hive.ql.exec.FetchOperator.pushRow(FetchOperator.java:414)
at org.apache.hadoop.hive.ql.exec.FetchTask.fetch(FetchTask.java:140)
at org.apache.hadoop.hive.ql.Driver.getResults(Driver.java:1670)
at org.apache.hive.service.cli.operation.SQLOperation.getNextRowSet(SQLOperation.java:347)
... 25 more
Caused by: java.lang.NullPointerException: null
at org.apache.carbondata.core.metadata.AbsoluteTableIdentifier.fromTablePath(AbsoluteTableIdentifier.java:84)
at org.apache.carbondata.hive.MapredCarbonInputFormat.populateCarbonTable(MapredCarbonInputFormat.java:105)
at org.apache.carbondata.hive.MapredCarbonInputFormat.getCarbonTable(MapredCarbonInputFormat.java:115)
at org.apache.carbondata.hive.MapredCarbonInputFormat.getQueryModel(MapredCarbonInputFormat.java:122)
at org.apache.carbondata.hive.MapredCarbonInputFormat.getRecordReader(MapredCarbonInputFormat.java:76)
at org.apache.hadoop.hive.ql.exec.FetchOperator$FetchInputFormatSplit.getRecordReader(FetchOperator.java:673)
at org.apache.hadoop.hive.ql.exec.FetchOperator.getRecordReader(FetchOperator.java:323)
at org.apache.hadoop.hive.ql.exec.FetchOperator.getNextRow(FetchOperator.java:445)
... 29 more

**Comment by cenyuhai** [comment]
I am also woking on it. Not just this issue make hive unable to use, but also store tableinfo in hive metastore

**Comment by cenyuhai** [comment]
can you try this pr: https://github.com/apache/carbondata/pull/1239

**Comment by anubhav100** [comment]
@cenyuhai there is one more problem we are not able to run hive example this PR also fixed that issue, can you please try running hive example from your PR I don't think that will work from your pr #1239

**Comment by cenyuhai** [comment]
@anubhav100  you are right, I also encounter your problem

**Comment by anubhav100** [comment]
@cenyuhai that problem is resolved by this pr

**Comment by anubhav100** [comment]
retest this please

**Comment by chenliang613** [comment]
retest this please

**Comment by ravipesala** [comment]
SDV Build Success with Spark 2.1, Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/203/

**Comment by chenliang613** [comment]
retest this please

**Comment by ravipesala** [comment]
SDV Build Success with Spark 2.1, Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/206/

**Comment by chenliang613** [comment]
retest this please

**Comment by chenliang613** [comment]
retest this please

**Comment by ravipesala** [comment]
SDV Build Success with Spark 2.1, Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/211/

File: integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonInputFormat.java [hunk 1]
```diff
@@ -94,20 +94,21 @@ private static void populateCarbonTable(Configuration configuration, String path
...
-    AbsoluteTableIdentifier absoluteTableIdentifier =
-        AbsoluteTableIdentifier.fromTablePath(validInputPath);
-    // read the schema file to get the absoluteTableIdentifier having the correct table id
-    // persisted in the schema
-    CarbonTable carbonTable = SchemaReader.readCarbonTableFromStore(absoluteTableIdentifier);
-    setTableInfo(configuration, carbonTable.getTableInfo());
-  }
+      AbsoluteTableIdentifier absoluteTableIdentifier =
+              AbsoluteTableIdentifier.fromTablePath(validInputPath);
+      // read the schema file to get the absoluteTableIdentifier having the correct table id
```
Comment (chenliang613) [inline integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonInputFormat.java]: This part code has some code style issue, please fix it.

**Comment by chenliang613** [comment]
@shivangi1015  as you provided steps, i verified it , it is working fine, thank you fixed this issue.
Please check CI, there are some code check style issues, please fix it, as below : 
/home/jenkins/jenkins-slave/workspace/carbondata-pr-spark-2.1/integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonInputFormat.java:104: error: 'member def type' have incorrect indentation level 6, expected level should be 4.
/home/jenkins/jenkins-slave/workspace/carbondata-pr-spark-2.1/integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonInputFormat.java:104: error: 'method def' child have incorrect indentation level 6, expected level should be 4.
/home/jenkins/jenkins-slave/workspace/carbondata-pr-spark-2.1/integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonInputFormat.java:108: error: 'member def type' have incorrect indentation level 6, expected level should be 4.
/home/jenkins/jenkins-slave/workspace/carbondata-pr-spark-2.1/integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonInputFormat.java:108: error: 'method def' child have incorrect indentation level 6, expected level should be 4.
/home/jenkins/jenkins-slave/workspace/carbondata-pr-spark-2.1/integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonInputFormat.java:109: error: 'method def' child have incorrect indentation level 6, expected level should be 4.
/home/jenkins/jenkins-slave/workspace/carbondata-pr-spark-2.1/integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonInputFormat.java:110: error: 'method def' child have incorrect indentation level 6, expected level should be 4.
/home/jenkins/jenkins-slave/workspace/carbondata-pr-spark-2.1/integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonInputFormat.java:111: error: 'method def rcurly' have incorrect indentation level 4, expected level should be 2.

**Comment by chenliang613** [comment]
LGTM

**Comment by ravipesala** [comment]
SDV Build Success with Spark 2.1, Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/215/

**Pull request closed without merging**